### DISCUSSION TITLE
What is the proper way for the socket.io server to know the client actually receives its emit message?

### DISCUSSION THREAD

**Discussion by qiulang** [socketio/socket.io#3846] [Q&A] [answered]
What is the proper way for the socket.io server to know the client actually receives its emit message? 

If ack callack is called I assume the client has received my message, if ack is not called I assume the client did not. But how do I know ack was not called by client, e.g. how long should I wait before I  can say the client failed to call ack ?

socket.emit('msg', (message, ack) => {
     // How do I know ack was not called by client ?
}) 

Or is there other way to know whether the client actually receives the emit message?

**Comment by qiulang** [comment, marked as answer]
OK I found https://socket.io/docs/v3/emitting-events/ has an acknowledgement timeout example. I am using 2.3 so I didn't notice the document when I asked the question.

So does this mean timeout without ack the proper way to know client doesn't receive the message ?

**Reply by darrachequesne** [reply]
> So does this mean timeout without ack the proper way to know client doesn't receive the message ?

Yes it is.

Please note that it is a bit more complex in the other direction (client to server), because the event may be buffered when the client is offline (in that case, you would need to remove the message from the buffer in the timeout handler).

Documentation: https://socket.io/docs/v3/client-offline-behavior/#Buffered-events

**Reply by qiulang** [reply]
Thanks for reply my question!
But what is recommended value for timeout ? In the example it is 1 second, or it is hard to recommend a good one?

**Reply by darrachequesne** [reply]
Yes, it will depend on your use case, you may want to display a spinner or a loading bar to the user, it's up to you.

1 second might be a bit low though.

**Reply by qiulang** [reply]
Before I knew this timeout way I actually designed a homemade method,

1. record the messages before I emit them.
2. use ack to delete the message, so the messages are not deleted will be regarded as failed ones.
3. use a timer to check the failed ones regularly. But how often should I check is a problem.

Do you think this method can be another way ?

Thanks.

**Reply by darrachequesne** [reply]
You can attach a timeout attribute (`Date.now() + 5000`, for example) to each recorded message, and periodically check if this timeout is reached. That should work too :+1:
